clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java

```diff
@@ -0,0 +1,212 @@
+                    serverNonce = formatter.secureRandomString();
+                    try {
+                        String saslName = clientFirstMessage.saslName();
+                        this.username = formatter.username(saslName);
+                        NameCallback nameCallback = new NameCallback("username", username);
+                        ScramCredentialCallback credentialCallback = new ScramCredentialCallback();
+                        callbackHandler.handle(new Callback[]{nameCallback, credentialCallback});
+                        this.scramCredential = credentialCallback.scramCredential();
+                        if (scramCredential == null)
+                            throw new SaslException("Authentication failed: Invalid user credentials");
+                        this.serverFirstMessage = new ServerFirstMessage(clientFirstMessage.nonce(),
+                                serverNonce,
+                                scramCredential.salt(),
+                                scramCredential.iterations());
+                        setState(State.RECEIVE_CLIENT_FINAL_MESSAGE);
+                        return serverFirstMessage.toBytes();
+                    } catch (IOException | NumberFormatException | UnsupportedCallbackException e) {
+                        setState(State.FAILED);
+                        throw new SaslException("Authentication failed: Credentials could not be obtained", e);
+                    }
+
+                case RECEIVE_CLIENT_FINAL_MESSAGE:
+                    try {
+                        ClientFinalMessage clientFinalMessage = new ClientFinalMessage(response);
+                        verifyClientProof(clientFinalMessage);
+                        byte[] serverKey = scramCredential.serverKey();
+                        byte[] serverSignature = formatter.serverSignature(serverKey, clientFirstMessage, serverFirstMessage, clientFinalMessage);
+                        ServerFinalMessage serverFinalMessage = new ServerFinalMessage(null, serverSignature);
+                        setState(State.COMPLETE);
+                        return serverFinalMessage.toBytes();
+                    } catch (InvalidKeyException e) {
+                        setState(State.FAILED);
+                        throw new SaslException("Authentication failed: Invalid client final message", e);
+                    }
+
+                default:
+                    throw new IllegalSaslStateException("Unexpected challenge in Sasl server state " + state);
+            }
+        } catch (SaslException e) {
+            setState(State.FAILED);
+            throw e;
+        }
+    }
+
+    @Override
+    public String getAuthorizationID() {
+        if (!isComplete())
+            throw new IllegalStateException("Authentication exchange has not completed");
+        String authzId = clientFirstMessage.authorizationId();
+        return authzId == null || authzId.length() == 0 ? username : authzId;
```

Comment:
It seems that the client always leaves authorizationId empty in the first message. So, we could just return username here?